apiserver/facades/client/bundle/bundle.go

```diff
@@ -150,11 +160,149 @@ func (b *BundleAPI) GetChanges(args params.BundleChangesParams) (params.BundleCh
+	return params.StringResult{}, common.ServerError(failErr)
+}
+
 // ExportBundle exports the current model configuration as bundle.
 func (b *BundleAPI) ExportBundle() (params.StringResult, error) {
-	return params.StringResult{}, nil
+	if err := b.checkCanRead(); err != nil {
+		return fail(err)
+	}
+
+	var exportConfig state.ExportConfig
+
+	b.backend.SetExportconfig(exportConfig)
+	logger.Criticalf("XXXXXXXXXXXXXXXXX...........SkipStatusHistory: %v ", exportConfig.SkipStatusHistory)
+	model, err := b.backend.ExportPartial(exportConfig)
+	if err != nil {
+		return fail(err)
+	}
+
+	// Fill it in charm.BundleData datastructure.
+	bundleData, err := b.FillBundleData(model)
+	if err != nil {
+		return fail(err)
+	}
+
+	bytes, err := yaml.Marshal(bundleData)
+	if err != nil {
+		return fail(err)
+	}
+
+	return params.StringResult{
+		Result: string(bytes),
+	}, nil
 }
 
 // Mask the new method from V1 API.
 // ExportBundle is not in V1 API.
 func (u *APIv1) ExportBundle() (_, _ struct{}) { return }
+
+// FillBundle fills the bundledata datastructure required for the exportBundle.
+func (b *BundleAPI) FillBundleData(model description.Model) (*charm.BundleData, error) {
+	cfg := model.Config()
+	var defaultSeries string
+	for key, _ := range cfg {
+		if key == "default-series" {
+			defaultSeries = fmt.Sprintf("%v", cfg[key])
+		}
+	}
+
+	data := &charm.BundleData{
+		Series:       defaultSeries,
+		Applications: make(map[string]*charm.ApplicationSpec),
+		Relations:    [][]string{},
+		Machines:     make(map[string]*charm.MachineSpec),
+	}
+
+	if len(model.Applications()) == 0 {
+		return &charm.BundleData{}, errors.Errorf("nothing to export as there is no application found.")
+	}
+	for _, application := range model.Applications() {
+		var newApplication *charm.ApplicationSpec
+		if application.Subordinate() {
+			newApplication = &charm.ApplicationSpec{
+				Charm:            application.CharmURL(),
+				Series:           application.Series(),
+				Expose:           application.Exposed(),
+				Options:          application.CharmConfig(),
+				Annotations:      application.Annotations(),
+				EndpointBindings: application.EndpointBindings(),
+			}
+		} else {
+			ut := []string{}
+			for _, unit := range application.Units() {
+				ut = append(ut, unit.Machine().Id())
+			}
+
+			newApplication = &charm.ApplicationSpec{
+				Charm:            application.CharmURL(),
+				Series:           application.Series(),
+				NumUnits:         len(application.Units()),
+				To:               ut,
+				Expose:           application.Exposed(),
+				Options:          application.CharmConfig(),
+				Annotations:      application.Annotations(),
+				EndpointBindings: application.EndpointBindings(),
+			}
+		}
+
+		data.Applications[application.Name()] = newApplication
+	}
+
+	for _, machine := range model.Machines() {
+		result := b.constraints(machine.Constraints())
+
+		constraints := []string{"arch=" + *result.Arch,
```

Comment:
We do not want to create all this if thher are not specified...

I think you want your b.constraints(...) to return this collection populating only the parameters that were specified...